In [1]:
import pandas as pd
from pandas import DataFrame

## Load police units from current database

In [3]:
df_database_units = DataFrame.from_records(PoliceUnit.objects.all().values('unit_name', 'description'))
df_database_units

description unit_name
0                      Area 0       000
1                District 001       001
2                District 002       002
3                District 003       003
4                District 004       004
5                District 005       005
6                District 006       006
7                District 007       007
8                District 008       008
9                District 009       009
10               District 010       010
11               District 011       011
12               District 012       012
13               District 013       013
14               District 014       014
15               District 015       015
16               District 016       016
17               District 017       017
18               District 018       018
19               District 019       019
20               District 020       020
21               District 021       021
22               District 022       022
23               District 023       023
24               District 024       024
25               District 025       025
26    Executive Officers Unit       026
27          Caps Crt Advocacy       032
28         Communications C/G       043
29   Recruit Training Section       044
..                        ...       ...
233                      None       912
234             Patrol Area 2       932
235      Cops Grant Recr Trng       944
236            Maint Only Omi       990
237      Area 2 Pat Narc Prog       062
238   Midway Airport Security       149
239      Area 1 Pat Narc Prog       061
240                      None       054
241      Youth Division Area6       076
242      Area 6 Pat Narc Prog       066
243                      None       056
244                      None       194
245      Area 3 Pat Narc Prog       063
246      Area 5 Pat Narc Prog       065
247                      None       064
248                      None       041
249                      None       052
250                      None       040
251                      None       661
252                      None       028
253                      None       271
254                      None       274
255                      None       047
256                      None       093
257                      None       272
258                      None       031
259                      None       275
260                      None       095
261                      None       185
262                      None       273

[263 rows x 2 columns]

** We get the list of unit name from our current database **

In [29]:
exist_units = set(df_database_units['unit_name'])

## Load cleaned unit history from cleaned data

In [5]:
df_all_sworn_units = pd.read_csv('../cleaned_data/unit-history/all-sworn-units.csv')
df_all_sworn_units

all_sworn_units_ID  Unit  Start.Date    End.Date    UID
0                    12735     3  2016-06-05         NaN      1
1                    12735    14  2007-03-29  2007-03-28      1
2                    12735    14  2007-03-29  2016-06-04      1
3                    16413    15  2007-03-29  2007-03-28      2
4                    16413    15  2007-03-29         NaN      2
5                     4767     1  1976-08-19  1976-08-20      3
6                     4767     4  1971-01-16  1976-08-18      3
7                     4767    22  1970-06-15  1971-01-15      3
8                     4767   543  1976-08-21         NaN      3
9                     1207    11  1996-09-12  2005-05-25      4
10                    1207    19  1995-11-09  1996-09-11      4
11                    1207    20  2005-05-26         NaN      4
12                    1207    44  1994-12-05  1995-11-08      4
13                    3134    16  1971-07-22  1973-06-15      5
14                    3134    18  1969-08-16  1971-07-21      5
15                    3134    22  1969-01-06  1969-08-15      5
16                    3134    65  1973-06-16  1973-07-24      5
17                    3134   640  1995-04-27         NaN      5
18                    3134   652  1973-07-25  1995-04-26      5
19                    3146    13  1976-06-22  1980-01-09      6
20                    3146    20  1971-08-08  1976-06-21      6
21                    3146    24  1980-01-10  1980-01-11      6
22                    3146    25  1980-01-12         NaN      6
23                    3146    32  1955-04-16  1961-04-15      6
24                    3146    44  1954-10-16  1955-04-15      6
25                    3146    65  1961-12-16  1965-08-18      6
26                    3146    66  1965-08-19  1971-08-07      6
27                    3146   167  1961-04-16  1961-12-15      6
28                   28676    12  1997-11-06  2013-10-12      7
29                   28676    16  2013-10-13         NaN      7
...                    ...   ...         ...         ...    ...
115957               10638   151  1981-09-19         NaN  31996
115958               19761    44  2015-08-31         NaN  31997
115959                2859     3  2009-03-05  2011-01-05  31998
115960                2859     7  2011-01-06         NaN  31998
115961                2859    44  2007-09-24  2009-03-04  31998
115962                5902    44  2016-02-29         NaN  31999
115963               12455     7  1980-12-11  1990-09-12  32000
115964               12455     8  1990-09-13  1998-11-11  32000
115965               12455    11  1966-05-09  1980-12-10  32000
115966               12455   188  1998-11-12         NaN  32000
115967               23248     7  1997-10-09  1999-05-26  32001
115968               23248     7  1999-05-27  2004-09-15  32001
115969               23248     8  1999-04-01  2007-02-28  32001
115970               23248    17  2013-01-06  2014-01-04  32001
115971               23248    19  2007-03-01  1997-10-08  32001
115972               23248    19  2007-03-01  2010-02-03  32001
115973               23248    19  2014-01-05         NaN  32001
115974               23248    25  2012-10-14  2013-01-05  32001
115975               23248   153  2004-09-16  2007-02-28  32001
115976               23248   313  2010-02-04  2012-10-13  32001
115977               16450    44  2016-02-29         NaN  32002
115978                6037    13  1987-09-17  1987-10-14  32003
115979                6037    24  1987-10-15  2000-08-31  32003
115980                6037    44  1986-10-13  1987-09-16  32003
115981                6037   650  2000-09-01         NaN  32003
115982               19389    14  2004-11-11  2004-11-10  32004
115983               19389    14  2004-11-11         NaN  32004
115984                3057    20  1997-11-06  1997-11-05  32005
115985                3057    20  1997-11-06         NaN  32005
115986               27928   631  1990-01-01         NaN  32006

[115987 rows x 5 columns]

** We get the list of unit name from all sworn units **

In [6]:
units = set(df_all_sworn_units['Unit'])

** Now, we get the unit names from all sworn units that don't exist in our current database **

In [11]:
new_units = [unit for unit in units if str(unit).zfill(3) not in exist_units]
list(new_units)

[720, -999]

Let exclude `-999` Roman used this value to replace the empty value in his cleaned data. So, we have only one new police unit **`720`**

## Investigate on police unit 720

** Get UID of officer who have police unit 720 **

In [14]:
df_all_sworn_units[df_all_sworn_units['Unit']==720]

all_sworn_units_ID  Unit  Start.Date End.Date    UID
113089               21351   720  2016-10-09      NaN  31174

And we found out there is only one officer 31174 has police unit 720

** Reach out officer profiles who have police unit 720 **

In [22]:
df_officer_profiles = pd.read_csv('../cleaned_data/profiles/officer-profiles.csv')
df_officer_profiles[df_officer_profiles['UID']==31174].loc[:, ['UID', 'all_sworn_units_ID', 'unit_history_ID', 'First.Name', 'Last.Name', 'Current.Unit']]

UID  all_sworn_units_ID  unit_history_ID First.Name Last.Name  \
31173  31174             21351.0          30774.0   MICHELLE    WILSON   

       Current.Unit  
31173         120.0

** Find the unit history of the above officer **

In [34]:
df_all_sworn_units[df_all_sworn_units['UID']==31174].sort_values('Start.Date')

all_sworn_units_ID  Unit  Start.Date    End.Date    UID
113087               21351    44  1999-06-21  2000-05-24  31174
113084               21351     6  2000-05-25  2005-07-20  31174
113086               21351    11  2005-07-21  2005-11-09  31174
113085               21351     6  2005-11-10  2007-08-31  31174
113083               21351     3  2007-09-01  2014-04-26  31174
113088               21351   120  2014-04-27  2016-10-08  31174
113089               21351   720  2016-10-09         NaN  31174

We detected a inconsistent in data here. Take a look on the history of officer, we think that the current unit should be **`720`**, but in officer profiles cleaned data, it's **`120`**.